lightningd/lightningd.c

```diff
@@ -101,36 +179,51 @@ static const char *daemons[] = {
 	"lightning_openingd"
 };
 
-/* Check we can run them, and check their versions */
-void test_daemons(const struct lightningd *ld)
+/*~ Check we can run them, and check their versions */
+void test_subdaemons(const struct lightningd *ld)
 {
 	size_t i;
-	for (i = 0; i < ARRAY_SIZE(daemons); i++) {
+
+	/*~ CCAN's ARRAY_SIZE() should always be used on defined arrays: it will
+	 * fail to build if the argument is actually a pointer, not an array! */
+	for (i = 0; i < ARRAY_SIZE(subdaemons); i++) {
 		int outfd;
-		const char *dpath = path_join(tmpctx, ld->daemon_dir, daemons[i]);
+		/*~ CCAN's path module uses tal, so wants a context to allocate
+		 * from.  We have a magic context 'tmpctx' which is freed in
```

Comment:
this is very helpful for me especially knowing that it is freed in the eventloop. but I think it needs more elaboration. Also You mention the eventloop here for the first time. I guess the eventloop is a crucial design decission that would be important to understand?